aspnetcore/performance/blazor/webassembly-best-practices.md

```diff
@@ -0,0 +1,153 @@
+
+For example, a grid or list that renders hundreds of rows containing components is processor intensive to render. Consider virtualizing a grid or list layout so that only a subset of the components is rendered at any given time. For an example of component subset rendering, see the following components in the [Virtualization sample app (aspnet/samples GitHub repository)](https://github.com/aspnet/samples/tree/master/samples/aspnetcore/blazor/Virtualization):
+
+* `Virtualize` component ([Shared/Virtualize.razor](https://github.com/aspnet/samples/blob/master/samples/aspnetcore/blazor/Virtualization/Shared/Virtualize.cs)) &ndash; A component written in C# that implements <xref:Microsoft.AspNetCore.Components.ComponentBase> to render a set of weather data rows based on user scrolling.
+* `FetchData` component ([Pages/FetchData.razor](https://github.com/aspnet/samples/blob/master/samples/aspnetcore/blazor/Virtualization/Pages/FetchData.razor)) &ndash; Uses the `Virtualize` component to display 25 rows of weather data at a time.
+
+## Avoid JavaScript interop to marshal data
+
+In Blazor WebAssembly, a JS interop call must traverse the WebAssembly-JS boundary. Serializing and deserializing content across the two contexts creates processing overhead for the app. Frequent JS interop calls often adversely affects performance. To reduce the marshalling of data across the boundary, determine if the app can consolidate many small payloads into a single large payload to avoid the high volume of context switching between WebAssembly and JS.
+
+## Use System.Text.Json
+
+Blazor's JS interop implementation relies on <xref:System.Text.Json>, which is a high performance JSON serialization library with low memory allocation. Using <xref:System.Text.Json> doesn't result in additional app payload size over adding one or more alternate JSON libraries.
+
+For migration guidance, see [How to migrate from Newtonsoft.Json to System.Text.Json](/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to).
+
+## Use synchronous and unmarshalled JS interop APIs where appropriate
+
+Blazor WebAssembly offers two additional versions of <xref:Microsoft.JSInterop.IJSRuntime> over the single version available to Blazor Server apps:
+
+* <xref:Microsoft.JSInterop.IJSInProcessRuntime> allows invoking JS interop calls synchronously, which has less overhead than the asynchronous versions:
+
+  ```razor
+  @inject IJSRuntime JS
+
+  @code {
+      protected override void OnInitialized()
+      {
+          var jsInProcess = (IJSInProcessRuntime)JS;
+
+          var value = jsInProcess.Invoke<string>("jsInteropCall");
+      }
+  }
+  ```
+
+* <xref:Microsoft.JSInterop.WebAssembly.WebAssemblyJSRuntime> permits unmarshalled JS interop calls. While this approach has the least overhead, the JavaScript APIs required to interact with these APIs are currently undocumented and subject to breaking changes in future releases:
```

Comment:
It might benefit to put a warning for this (at least until they stabilize these APIs)